### PR TITLE
Improved scene detection

### PR DESCRIPTION
**Pull request by master-of-zen**
# Goals and motivation
Current fast scene detection in rav1e relatively slow to compared methods of scene detection, and have proclivity to show false results or don't detect scene changes where they appear

This pr reworks fast scene detection algorithm, making it faster, better, and more accurate

Achieved goals are:
* Faster decision making ( Both less and more efficient computations )
* More accurate scene detection, by adjusting threshold based on previous frames
* Frame downscale for faster decisions.

## Example of adaptive threshold not  cutting high motion segment which default static threshold method would cut
```
?  [SC-Detect] Frame 519: T=25.0 P=9.5 No cut
?  [SC-Detect] Frame 520: T=25.0 P=10.2 No cut
?  [SC-Detect] Frame 521: T=25.0 P=17.2 No cut

?  [SC-Detect] P: 29.0 [0.0, 1.8, 9.5, 10.2, 17.2] Cut: false
?  [SC-Detect] Frame 522: T=25.0 P=29.0 No cut

?  [SC-Detect] P: 33.6 [1.8, 9.5, 10.2, 17.2, 29.0] Cut: false
?  [SC-Detect] Frame 523: T=25.0 P=33.6 No cut

?  [SC-Detect] P: 34.0 [9.5, 10.2, 17.2, 29.0, 33.6] Cut: false
?  [SC-Detect] Frame 524: T=25.0 P=34.0 No cut

?  [SC-Detect] P: 26.4 [10.2, 17.2, 29.0, 33.6, 34.0] Cut: false
?  [SC-Detect] Frame 525: T=25.0 P=26.4 No cut

?  [SC-Detect] Frame 526: T=25.0 P=22.8 No cut
?  [SC-Detect] Frame 527: T=25.0 P=15.2 No cut
?  [SC-Detect] Frame 528: T=25.0 P=0.4 No cut
```
### 2160p
new | old 
:-------------------------:|:-------------------------:
| 

### 720p
new | old
:-------------------------:|:-------------------------:
||   |

### CODE REVIEW

**Review by vibhoothi — DISMISSED** [review]
Hi,

Thanks for the patch, 57 commits for this is nto easy to review, please squash the relevant commits, so reviewers can look it in better way

**Review by lu-zero — CHANGES_REQUESTED** [review]
Please squash and rebase on top of the current tree.

**Comment by shssoichiro** [comment]
My primary concern is that this seems to remove the scene flash detection. This was used for both fast and normal (cost-based) detection and is based on what x264 does to try and avoid an extra keyframe for a very short (5-frame-or-less) scene, preferring to just put one keyframe after the flash. In the past, it also helped to alleviate some false positives with pans being detected as scenecuts using the old fast method--I'd like to know if that issue has been resolved with the new fast algorithm?

**Comment by master-of-zen** [comment]
@shssoichiro 
I overlooked the flash detection for new algorithm and improvements should be made.

At this moment, new algorithm should handle pans just as fine as it use dequeue of scores to adjust threshold, which should also handle high motion / fade-in / fade-out

Making score dequeue bidirectional (compare score for current frame to scores before and after) should fix the issue flash detection, and should give free speed boost for all speeds as it reduce amount of computation

**Comment by lu-zero** [comment]
Can you please rebase this? :)

**Comment by master-of-zen** [comment]
It's wip) it will crash)

File: src/scenechange/mod.rs [hunk 1]
```diff
@@ -104,237 +136,289 @@ impl SceneChangeDetector {
...
-
-    self.is_key_frame(
-      frame_set[0].clone(),
-      frame_set[1].clone(),
-      input_frameno,
-      previous_keyframe,
-    )
-  }
+    // Initially fill score deque with forward frames
+    // ititiallization is different depending on frame set length
```
Comment (vibhoothiiaanand) [inline src/scenechange/mod.rs:140]:
```suggestion
    // Initialization is different depending on frame set length
```
Need to fix the typo+ maybe a rewording is a good idea, like merging both sentences? 
Initially fill score deque with forward frames based on frame_set length

File: src/scenechange/mod.rs [hunk 2]
```diff
@@ -104,237 +136,289 @@ impl SceneChangeDetector {
...
-      // and really wasteful for compression.
-      for frame in frameno..=(frameno + lookahead_distance as u64) {
-        self.excluded_frames.insert(frame);
+    if scenecut {
+      // Clear buffers and deque
+      self.frame_buffer.clear();
+      debug!("[SC-score-deque]{:.0?}", self.score_deque);
+      self.score_deque.clear();
+    } else {
+      // Keep score deque 5 + lookahead_size frames
```
Comment (vibhoothiiaanand) [inline src/scenechange/mod.rs:190]: I wonder how did we reach 5+ lookahead_size, would be nice to add in commit message or here, whichever is better.

On seeing further code, could see the mention of 5 logic, would be better if it is mentioned here than later.

File: src/scenechange/mod.rs [hunk 3]
```diff
@@ -104,237 +136,289 @@ impl SceneChangeDetector {
...
-      let delta = self.delta_in_planes(&frame1.planes[0], &frame2.planes[0]);
-      let threshold = self.threshold * len as u64;
-      ScenecutResult {
-        intra_cost: threshold as f64,
-        threshold: threshold as f64,
-        inter_cost: delta as f64,
-        has_scenecut: delta >= threshold,
-      }
+  ) -> ScenecutData {
+    let frame2_ref2 = Arc::clone(&frame2);
```
Comment (vibhoothiiaanand) [inline src/scenechange/mod.rs:318]: Is there frame2_ref1 or frame2_ref ?
It maybe slightly confusing if we have 2 at the end if there is no frame2_ref or frame2_ref1. Could simply keep ref itself, 
IIRC it was there earlier, itself, maybe @lu-zero could say if it is good idea to keep frame2_ref itself or not.

**Review by vibhoothiiaanand — COMMENTED** [review]
I did not check the code in-depth, was seeing the general implementation, I have some suggestions before landing 
+ The documentation for various things are good, but I felt the linebreaks are not consistent, sometimes happening after 20/30words, sometimes at 40/60, would be nice to be more consistent
+ The commit logs, would require a cleanup of commits to see by splitting into two/three, and giving a gist of the changes int eh commit messages, maybe could see the latest PR from @shssoichiro where he was giving a good explanation of things either in the commit log or the PR.

Apart from that It is good in shape generally. Thanks for the work.

@lu-zero would be nice if you can take a close-look in the rust part of APIs :)

File: src/scenechange/mod.rs (hunk 2 above)
Comment (shssoichiro) [reply src/scenechange/mod.rs:190]: IIRC 5 was chosen because it's the size of a frame pyramid. I think at one point there was at least one comment mentioning that, although I haven't touched this code in a while so my memory is fuzzy.

File: src/scenechange/mod.rs [hunk 4]
```diff
@@ -104,237 +136,288 @@ impl SceneChangeDetector {
...
       let delta_line = l1
         .iter()
         .zip(l2.iter())
         .map(|(&p1, &p2)| {
-          (i16::cast_from(p1) - i16::cast_from(p2)).abs() as u64
+          (i16::cast_from(p1) - i16::cast_from(p2)).abs() as u32
         })
-        .sum::<u64>();
+        .sum::<u32>();
       delta += delta_line;
```
Comment (barrbrain) [inline src/scenechange/mod.rs:387]: We need to up-cast to u64 here to accumulate the line sums. Otherwise, we will be limited to `2^9` pixels altogether, not just in each row.
```suggestion
      delta += delta_line as u64;
```

File: src/scenechange/mod.rs (hunk 1 above)
Comment (master-of-zen) [reply src/scenechange/mod.rs]: Fixed typos

File: src/scenechange/mod.rs [hunk 5]
```diff
@@ -81,18 +114,17 @@ impl SceneChangeDetector {
...
     &mut self, frame_set: &[Arc<Frame<T>>], input_frameno: u64,
     previous_keyframe: u64,
   ) -> bool {
+    // Use score deque for adaptive threshold for scene cut
+    // Declare score_deque offset based on lookahead  for scene change scores
+
     // Find the distance to the previous keyframe.
     let distance = input_frameno - previous_keyframe;
 
-    if frame_set.len() < 2 {
```
Comment (lu-zero) [inline src/scenechange/mod.rs:91]: why did you remove this check?       

I see a 
``` rust
 self.deque_offset = frame_set.len() - 2;
```

just below.

File: src/scenechange/mod.rs (hunk 3 above)
Comment (master-of-zen) [reply src/scenechange/mod.rs:317]: I don't know why it's done the way it is, I just refactored the code)

File: src/scenechange/mod.rs (hunk 5 above)
Comment (master-of-zen) [reply src/scenechange/mod.rs:91]: IIRC because scene change could handle 2 frames just fine.
I probably should add that back

Comment (master-of-zen) [reply src/scenechange/mod.rs:91]: Added back

File: src/scenechange/mod.rs (hunk 4 above)
Comment (master-of-zen) [reply src/scenechange/mod.rs]: Fixed

File: src/scenechange/mod.rs (hunk 2 above)
Comment (master-of-zen) [reply src/scenechange/mod.rs]: Improved comment

File: src/scenechange/mod.rs [hunk 6]
```diff
@@ -104,237 +140,288 @@ impl SceneChangeDetector {
...
+    input_frameno: u64, previous_keyframe: u64,
+  ) {
+    let result = if self.fast_mode {
+      self.fast_scenecut(frame1, frame2)
+    } else {
+      self.cost_scenecut(frame1, frame2, input_frameno, previous_keyframe)
+    };
+    self
+      .score_deque
+      .insert(0, (result.inter_cost as f64, result.intra_cost as f64));
```
Comment (barrbrain) [inline src/scenechange/mod.rs:232]: The threshold computation in `cost_scenecut()` is thrown out here.
```suggestion
      .insert(0, (result.inter_cost as f64, result.threshold as f64));
```

**Review by vibhoothi — COMMENTED** [review]
Further to add here to the PR from the #daala IRC discussion, 
On checking the BD-Rate[1], there seems regression of 8-14% BD-Rates on Speed-10, and also 1 frame displacement in the new scene-change mechanism[2], on lower speed levels tested, the BD-Rate remains neutral, while giving a boost in encoding/decoding time generally.

The subsequent frames in the decoded output have +1 change throughout, this happens in the case of a scene-change.
Things to try out, 
+ Examine the Output of new scene-change with different decoders, and extracting frame-by-frame(maybe with ffmpeg), and see the frames which shown the issue.
+ I did not check the code properly to find the +1 change of frame, it is happening to some clips which have scene-change at some levels, need to test at different levels/disabling some encoding features and seeing if it is an offset of some other encoding function, 
 
[1]: https://gist.github.com/vibhoothi/ff70287d120cf7dfe28dbbfa543e2f6c
[2]: https://beta.arewecompressedyet.com/analyzer/?maxFrames=30&decoder=https://mindfreeze.tk/~mindfreeze/inspect.js&decoderName=base_scene_change&file=https://mindfreeze.tk/~mindfreeze/scenechange/basescenechange.ivf&decoder=https://mindfreeze.tk/~mindfreeze/inspect.js&decoderName=new_scene_change&file=https://mindfreeze.tk/~mindfreeze/scenechange/newscenechange.ivf

File: src/scenechange/mod.rs (hunk 6 above)
Comment (master-of-zen) [reply src/scenechange/mod.rs]: Fixed

**Comment by lu-zero** [comment]
Please rebase the whole thing.

**Comment by master-of-zen** [comment]
# AWCY vimeo 10s corpus

## speed 6
[AWCY](https://beta.arewecompressedyet.com/?job=rav1e-master-vimeo-corpus-10s-8c064f4%402021-07-09T21%3A04&job=rav1e-newsc-vimeo-corpus-10s-44bfce4%402021-07-09T21%3A08%3A28.132Z)
 


## speed 9
[AWCY](https://beta.arewecompressedyet.com/?job=rav1e_master_s9%402021-07-10T22%3A39%3A15.575Z&job=av1e_new_scene_detection_s9%402021-07-10T22%3A38%3A43.915Z)

image


## speed 10 
[AWCY](https://beta.arewecompressedyet.com/?job=rav1e-master-s10-vimeo-corpus-10s-44bfce4%402021-07-09T21%3A08%3A48.497Z&job=rav1e-newsc-s10-vimeo-corpus-10s-44bfce4%402021-07-09T21%3A10%3A52.337Z)


## speed 10 with default scene detection and new algo
[AWCY](https://beta.arewecompressedyet.com/?job=rav1e-newsc-s10-vimeo-corpus-10s-44bfce4%402021-07-09T21%3A10%3A52.337Z&job=rav1e_new_scene_detection_only_default_scenechange_s10%402021-07-11T00%3A24%3A01.276Z)

**Review by lu-zero — APPROVED** [review]
It is fine for me if it is fine for @shssoichiro .

**Comment by shssoichiro** [comment]
It is good for me as well
